tests/unit/s2n_server_key_share_extension_test.c

```diff
@@ -0,0 +1,356 @@
+        EXPECT_FAILURE(s2n_extensions_server_key_share_send(conn, extension_stuffer));
+
+        conn->secure.client_ecc_params[0].negotiated_curve = &s2n_ecc_supported_curves[0];
+        EXPECT_FAILURE(s2n_extensions_server_key_share_send(conn, extension_stuffer));
+
+        EXPECT_SUCCESS(s2n_ecc_generate_ephemeral_key(&conn->secure.client_ecc_params[0]));
+        EXPECT_SUCCESS(s2n_extensions_server_key_share_send(conn, extension_stuffer));
+
+        conn->secure.client_ecc_params[0].negotiated_curve = &s2n_ecc_supported_curves[1];
+        EXPECT_FAILURE(s2n_extensions_server_key_share_send(conn, extension_stuffer));
+
+        EXPECT_SUCCESS(s2n_ecc_params_free(&conn->secure.server_ecc_params));
+        EXPECT_SUCCESS(s2n_connection_free(conn));
+    }
+
+    /* Test s2n_extensions_server_key_share_recv with supported curves */
+    {
+        for (int i = 0; i < S2N_ECC_SUPPORTED_CURVES_COUNT; i++) {
+            struct s2n_connection *server_send_conn;
+            struct s2n_connection *client_recv_conn;
+
+            EXPECT_NOT_NULL(server_send_conn = s2n_connection_new(S2N_SERVER));
+            EXPECT_NOT_NULL(client_recv_conn = s2n_connection_new(S2N_CLIENT));
+
+            struct s2n_stuffer* extension_stuffer = &server_send_conn->handshake.io;
+
+            server_send_conn->secure.server_ecc_params.negotiated_curve = &s2n_ecc_supported_curves[i];
+            server_send_conn->secure.client_ecc_params[i].negotiated_curve = &s2n_ecc_supported_curves[i];
+            EXPECT_SUCCESS(s2n_ecc_generate_ephemeral_key(&server_send_conn->secure.client_ecc_params[i]));
+
+            EXPECT_SUCCESS(s2n_extensions_server_key_share_send(server_send_conn, extension_stuffer));
+
+            S2N_STUFFER_READ_EXPECT_EQUAL(extension_stuffer, TLS_EXTENSION_KEY_SHARE, uint16);
+            S2N_STUFFER_READ_EXPECT_EQUAL(extension_stuffer, s2n_extensions_server_key_share_send_size(server_send_conn) - 2, uint16);
+
+            client_recv_conn->secure.client_ecc_params[i].negotiated_curve = &s2n_ecc_supported_curves[i];
+            EXPECT_SUCCESS(s2n_ecc_generate_ephemeral_key(&client_recv_conn->secure.client_ecc_params[i]));
+
+            /* Parse key share */
+            EXPECT_SUCCESS(s2n_extensions_server_key_share_recv(client_recv_conn, extension_stuffer));
+            EXPECT_EQUAL(s2n_stuffer_data_available(extension_stuffer), 0);
+
+            EXPECT_EQUAL(server_send_conn->secure.server_ecc_params.negotiated_curve->iana_id, client_recv_conn->secure.server_ecc_params.negotiated_curve->iana_id);
+            EXPECT_EQUAL(server_send_conn->secure.server_ecc_params.negotiated_curve, &s2n_ecc_supported_curves[i]);
+
+            EXPECT_SUCCESS(s2n_connection_free(server_send_conn));
+            EXPECT_SUCCESS(s2n_connection_free(client_recv_conn));
+        }
+    }
+
+    /* Test s2n_extensions_server_key_share_recv with various sample payloads */
```

Comment:
I [still don't think](https://github.com/awslabs/s2n/pull/1151#discussion_r320459833) these tests belong in this class. You're not testing s2n_extensions_server_key_share, you're testing s2n_ecc functions. While testing key serialization is definitely worthwhile, s2n_extensions_server_key_share is not where the logic that cares about the key payloads is.

Reply:
I'm not sure if there's a concern here - these are valid key_share extensions. Since key_share_recv() internally parses a key, from a blackbox approach of testing this function, we should be able to observe whether it parses a key share extension that is valid, or invalid (eg. if there's a corrupted key, does this function handle the error correctly).

I'm not sure if there are better ways to test this, other than
a) you encode a payload using key_share_send() (we are doing eariler)
b) you place some bytes into a stuffer (done here)
c) fuzzing